compiler/crates/relay-lsp/src/rename/mod.rs

```diff
@@ -0,0 +1,416 @@
+        let new_file_name = new_path.file_stem().unwrap().to_str().unwrap();
+
+        if old_file_name == new_file_name {
+            continue;
+        }
+
+        let full_text = get_file_contents(&old_path).unwrap();
+
+        let embedded_sources = extract_graphql::extract(&full_text);
+        if embedded_sources.is_empty() {
+            continue;
+        }
+
+        let program = &state.get_program(&state.extract_project_name_from_url(&old_file_uri)?)?;
+        let root_dir = &state.root_dir();
+
+        let mut index = 0;
+        for embedded_source in &embedded_sources {
+            // todo: do the rusty way
+            let _ = match embedded_source {
+                JavaScriptSourceFeature::GraphQL(graphql_source) => {
+                    let source_location_key =
+                        SourceLocationKey::embedded(new_file_uri.as_ref(), index);
+
+                    let text_source = graphql_source.text_source();
+                    let document = parse_executable_with_error_recovery(
+                        &text_source.text,
+                        source_location_key,
+                    )
+                    .item;
+
+                    for definition in &document.definitions {
+                        let changes = match definition {
+                            graphql_syntax::ExecutableDefinition::Fragment(frag_def) => {
+                                let frag_name = frag_def.name.value;
+                                let old_frag_name = frag_name.to_string();
+                                let new_frag_name =
+                                    old_frag_name.replace(old_file_name, new_file_name);
+
+                                rename_fragment(frag_name, new_frag_name, program, root_dir)
+                            }
+                            graphql_syntax::ExecutableDefinition::Operation(op_def) => {
+                                let operation_name_identifier = op_def.name.unwrap();
+                                let old_operation_name = operation_name_identifier.to_string();
+                                let new_operation_name =
+                                    old_operation_name.replace(old_file_name, new_file_name);
+
+                                let name_range =
+                                    text_source.to_span_range(operation_name_identifier.span);
+
+                                let location = Location::new(old_file_uri.clone(), name_range);
+
+                                rename_operation(new_operation_name, location)
+                            }
+                        };
+
+                        merge_text_changes(&mut rename_changes, changes);
+                    }
+
+                    Ok(())
+                }
+                _ => Err(LSPRuntimeError::ExpectedError),
+            };
+
+            index += 1;
+        }
+    }
+
+    Ok(Some(WorkspaceEdit {
+        changes: Some(rename_changes),
+        ..Default::default()
+    }))
```

Comment:
I'm really unhappy with this code, but I'm just hitting the ceiling of my Rust knowledge. I tried to make it more functional by using iterators, but I struggled with `collect`, ignore errors while iterating and in general how to avoid the many `.unwrap()` calls.
I would appreciate input from some Rust veterans 🦀 Also feel free to just make the changes yourself in the code!

Reply:
I've already made a couple of small improvements locally today, but will have to rebase since you changed a fair amount now. I'll walk you through it tomorrow.